Zend/zend_compile.c

```diff
@@ -5154,6 +5154,249 @@ void zend_compile_switch(zend_ast *ast) /* {{{ */
+		for (uint32_t j = 0; j < conds->children; j++) {
+			zend_ast **cond_ast = &conds->child[j];
+
+			zend_eval_const_expr(cond_ast);
+			if ((*cond_ast)->kind != ZEND_AST_ZVAL) {
+				/* Non-constant condition */
+				return IS_UNDEF;
+			}
+
+			zval *cond_zv = zend_ast_get_zval(*cond_ast);
+			if (Z_TYPE_P(cond_zv) != IS_LONG && Z_TYPE_P(cond_zv) != IS_STRING) {
+				/* We only optimize switched on integers and strings */
+				return IS_UNDEF;
+			}
+
+			if (common_type == IS_UNDEF) {
+				common_type = Z_TYPE_P(cond_zv);
+			} else if (common_type != Z_TYPE_P(cond_zv)) {
+				/* Non-uniform condition types */
+				return IS_UNDEF;
+			}
+		}
+	}
+
+	return common_type;
+}
+
+void zend_compile_match(znode *result, zend_ast *ast) /* {{{ */
+{
+	zend_ast *expr_ast = ast->child[0];
+	zend_ast_list *arms = zend_ast_get_list(ast->child[1]);
+
+	if (arms->children == 0) {
+		result->op_type = IS_CONST;
+		ZVAL_NULL(&result->u.constant);
+	}
+
+	zend_bool has_default_arm = 0;
+
+	uint32_t opnum_match = (uint32_t)-1;
+
+	znode expr_node;
+	zend_compile_expr(&expr_node, expr_ast);
+
+	zend_begin_loop(ZEND_FREE, &expr_node, 0);
+
+	znode case_node;
+	case_node.op_type = IS_TMP_VAR;
+	case_node.u.op.var = get_temporary_variable();
+
+	uint32_t num_conds = count_match_conds(arms);
+	zend_uchar jumptable_type = determine_match_jumptable_type(arms);
+	zend_bool uses_jumptable = jumptable_type != IS_UNDEF && should_use_jumptable(num_conds, jumptable_type);
+	HashTable *jumptable = NULL;
+	uint32_t *jmpnz_opnums = NULL;
+
+	for (uint32_t i = 0; i < arms->children; ++i) {
+		zend_ast *arm_ast = arms->child[i];
+
+		if (!arm_ast->child[0]) {
+			if (has_default_arm) {
+				CG(zend_lineno) = arm_ast->lineno;
+				zend_error_noreturn(E_COMPILE_ERROR,
+					"Match expressions may only contain one default arm");
+			}
+			has_default_arm = 1;
+		}
+	}
+
+	if (uses_jumptable) {
+		znode jumptable_op;
+
+		ALLOC_HASHTABLE(jumptable);
+		zend_hash_init(jumptable, arms->children, NULL, NULL, 0);
+		jumptable_op.op_type = IS_CONST;
+		ZVAL_ARR(&jumptable_op.u.constant, jumptable);
+
+		zend_op *opline = zend_emit_op(NULL,
+			jumptable_type == IS_LONG ? ZEND_MATCH_LONG : ZEND_MATCH_STRING,
+			&expr_node, &jumptable_op);
+		if (opline->op1_type == IS_CONST) {
+			Z_TRY_ADDREF_P(CT_CONSTANT(opline->op1));
+		}
+		opnum_match = opline - CG(active_op_array)->opcodes;
+	} else {
+		jmpnz_opnums = safe_emalloc(sizeof(uint32_t), num_conds, 0);
+		int cond_count = 0;
+		for (uint32_t i = 0; i < arms->children; ++i) {
+			zend_ast *arm_ast = arms->child[i];
+
+			if (!arm_ast->child[0]) {
+				continue;
+			}
+
+			zend_ast_list *conds = zend_ast_get_list(arm_ast->child[0]);
+			for (uint32_t j = 0; j < conds->children; j++) {
+				zend_ast *cond_ast = conds->child[j];
+
+				znode cond_node;
+				zend_compile_expr(&cond_node, cond_ast);
+
+				zend_op *opline = zend_emit_op(NULL, ZEND_IS_IDENTICAL, &expr_node, &cond_node);
```

Comment:
This is going to multi-free expr_node. Please add some tests that don't use JT and have an allocated match operand.

Reply:
I'm assuming this will need a new opcode?

Reply:
Yes, that's right. Similar to have we have ZEND_CASE as a switch-variant of IS_EQUAL.